Revamp UI interactions

jcollins1983 opened the pull request:
This PR addresses Issue #177 and Issue #213.

I've done a similar thing to the switching module and put the new _ui.py module at the fixate level and have only exposed the public functions.

To avoid having 2 different sets of functions in `cmd_line.py` and `ui_gui_qt.py` I moved the (duplicated) logic out of the UI files into ui.py and kept the original behaviour, tests were updated to deal with this change.

Also added the ability to have colour on `user_info_important` the line of `!` defaults to red, which will hopefully be a bit more eye catching.

Have tested each of the functions manually. But should try on a real script or 2 before this gets merged if it does.

jcollins1983 committed:
make reformat_text private

jcollins1983 committed:
beginnings of (hopefully) improving the UI logic

jcollins1983 committed:
switch to the more descriptive UserInputError from the Fixate exceptions module

jcollins1983 committed:
switch to new user_serial from _ui.py

jcollins1983 committed:
add user_yes_no and _user_abort_retry

jcollins1983 committed:
add _user_choices_

jcollins1983 committed:
add space after choices string

jcollins1983 committed:
make get_user_input private

jcollins1983 committed:
gui for user choices done

jcollins1983 committed:
add some colour to the user_info_important call

jcollins1983 committed:
make a bit easier to follow

jcollins1983 committed:
add user_info and user_info_important

jcollins1983 committed:
add user_ok

jcollins1983 committed:
add user_action

jcollins1983 committed:
add some missing type hints

jcollins1983 committed:
add _user_image so indicate that an image would have been displayed

jcollins1983 committed:
make information standout more

jcollins1983 committed:
fix typo

jcollins1983 committed:
add user image and gif functionality

jcollins1983 committed:
add doc strings

jcollins1983 committed:
include message RE use of GIFs in ccommand line

jcollins1983 committed:
add post sequence display functions

jcollins1983 committed:
move logic out of cmd and qt UIs into ui.py

jcollins1983 committed:
adjust tests to account for movement of logic from ui to ui controller (ui.py)

jcollins1983 committed:
fix a whoopsie

jcollins1983 committed:
fix docstring

jcollins1983 committed:
probably don't need anything other than int or string for serial numbers

jcollins1983 committed:
add tests for new _ui.py

jcollins1983 committed:
fix tests

jcollins1983 committed:
add a bit more flexibility for the user_info_important colours in the GUI

jcollins1983 committed:
add test to ensure Issue #213 is not repeated

jcollins1983 commented:
Looks like only the `readthedocs` build is done, no other tests? 🤔
<img width="828" alt="image" src="https://github.com/user-attachments/assets/57d50870-423a-4b76-8fa9-2fe3a002de26">

jcollins1983 commented:
This'll be why
<img width="295" alt="image" src="https://github.com/user-attachments/assets/c51e325c-a3d1-40c2-b468-672029f1bdea">

jcollins1983 commented:
> This'll be why <img alt="image" width="295" src="https://private-user-images.githubusercontent.com/28747582/363471834-c51e325c-a3d1-40c2-b468-672029f1bdea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUxNzc4MzMsIm5iZiI6MTcyNTE3NzUzMywicGF0aCI6Ii8yODc0NzU4Mi8zNjM0NzE4MzQtYzUxZTMyNWMtYTNkMS00MGMyLWI0NjgtNjcyMDI5ZjFiZGVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAxVDA3NTg1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2OTk4OGNiYWE2NGUxMmE1NDg4NTQwMzVjMzFhMDM5Njg5MTllMTNkNDVkZmQzNWNhZjljYjQ0NTQ0Njg4OTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wwQdcRF4we6nkTuhP6DZu4jfIgJA4HXQzAAd0t8VnFo">

I'll do another targeted branch to fix this.

clint-lawrence commented:
I haven't looked at the details here yet. General idea look fine.

This is what I'm thinking - there is probably a fair bit more we can/should do to improve. For example, tighten up typing to be more specific instead of many parameters/return values being strings. I don't want to get this improvement bogged down in trying to fix everything, but I don't want to shoot ourselves in the foot and complicate future improvements.

In any case, I'll get in and review this as is. And hopefully none of that is an issue.

jcollins1983 committed:
Merge branch 'PyFixate:main' into revamp-ui

jcollins1983 committed:
fix typo in comment

daniel-montanari commented on src/fixate/ui_gui_qt/ui_gui_qt.py:
```diff
@@ -839,23 +775,32 @@ def _topic_UI_display(self, msg):
 
+        txt_fill = "!" * wrapper.width
+
+        # Hisotry window
```
how could you

jcollins1983 replied:
It took me a second to see what you were talking about... fixed.

daniel-montanari commented on src/fixate/ui_gui_qt/ui_gui_qt.py:
```diff
@@ -738,96 +738,32 @@ def _topic_UI_action(self, msg, callback_obj):
+
+    def _topic_UI_req_choices(self, msg, q, choices):
         if self.closing:
+            # I don't think the result of this code ever gets used, nothing looking for "ABORT_FORCE"
+            # unpacks from a tuple. This can probably be deleted.
```
2 things in main look for this, I really don't understand what these are doing though. Doesn't make sense to me that we would inidicate a forced abort during a user input function

jcollins1983 replied:
It's possible that that was added after I started this work. The blame says it's a similar age. I've removed the comments and adjusted the queue additions, we're not using the ("Result", "Message") format anymore.

daniel-montanari commented on src/fixate/ui_gui_qt/ui_gui_qt.py:
Do you want to give `reformat_text` the same treatment you did in the cli?

jcollins1983 replied:
Sure, doesn't hurt to be consistent, so I don't recall seeing the same issues in the GUI (but I wouldn't trust my memory from back then)

daniel-montanari replied:
look good

daniel-montanari reviewed (commented):
Looks good overall. Needs the 3.11 features fixed for 3.8.

jcollins1983 committed:
Merge branch 'main' into revamp-ui

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+
+from typing import Callable, Any
+from queue import Queue, Empty
+from enum import StrEnum
```
Now that we've dropped python prior to 3.12, this won't be an issue.

jcollins1983 committed:
Merge branch 'main' into revamp-ui

jcollins1983 committed:
remove the 'Result' part of the q.put

jcollins1983 committed:
fix the things that were getting in the way of the UI tests

jcollins1983 committed:
make mypy happy

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+        return self.error_msg
+
+
+class UiColour(StrEnum):
```
Now that we've dropped python prior to 3.12, this won't be an issue.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+        return self.func(resp)
+
+    def __str__(self) -> str:
+        return self.error_msg
```
Will be fixed in an incoming commit.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    GREY = "grey"
+
+
+def _user_request_input(msg: str):
```
Fixed in incoming commit. I would have thought that mypy would flag an untyped return.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+def user_serial(
+    msg: str,
+    validator: Validator = _ten_digit_int_serial_v,
+    return_type: int | str = int,
```
typing.Type is [deprecated](https://docs.python.org/3.13/library/typing.html#typing.Type). [type](https://docs.python.org/3.13/library/functions.html#type) is the thing to use now.

Mypy wasn't happy with how this was done anyway.

Also, this is the problem with starting something then not touching it for over a year. It's probably 6 of one half a dozen of the other in terms of doing it this way or always returning a string then manually converting later.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    validator: Validator = _ten_digit_int_serial_v,
+    return_type: int | str = int,
+    attempts: int = 5,
+) -> Any:
```
I've fixed this, but using the python 3.10+ syntax

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    return resp
+
+
+def _choice_from_response(choices: tuple, resp: str) -> str | bool:
```
Fixed in incoming commit.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    raise UserInputError("User failed to enter the correct format serial number")
+
+
+def _user_req_choices(msg: str, choices: tuple):
```
Fixed in incoming commit

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    try:
+        while True:
+            try:
+                callback_obj.user_cancel_queue.get_nowait()
```
I'm not getting an error when I run mypy...

daniel-montanari replied:
Looks like this setting isn't turned on for this file. I might've had everything at strict when I wrote this.

I think for this PR I would keep as is, but we probably need to make it more obvious in the project readme how to set these sorts of things up when writing new modules. 
The general intent is over time we fix up errors and add in stricter options.

jcollins1983 replied:
Ok, agree. I've added an assert as the developer making tweaks or additions to the UI should be looking after the assignment of `user_cancel_queue` and `target_finished_callback`. Also added typing for these.

daniel-montanari replied:
Looks good.

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+        bool: True if the action is finished, False otherwise
+    """
+    # UserActionCallback is used to handle the cancellation of the action either by the user or by the action itself
+    class UserActionCallback:
```
I'm not sure I follow. This is defined in the common module that both cmd and gui use.

daniel-montanari replied:
I think what I meant was in _topic_UI_action and _user_action to add some typing info on what callback_obj is.

jcollins1983 replied:
In all cases it should be a `UserActionCallback`, but this is defined within the `user_action` method, I don't see an easy way of typing this in the calling functions. I have noticed though that the docstrings for the cmdline was incorrect, so I've updated that.

daniel-montanari replied:
I think thats what I was getting at.
Other modules use the callback so it probably shouldn't live inside `user_action`. Since they aren't checked by mypy there isn't too much benefit to the type hinting I was originally talking about.

jcollins1983 replied:
The callback is called from other modules, that's what callbacks are designed for. The callback is specifically for handling cancellation of `user_action` nothing else, so I see no reason for it _not_ to be where it is.

daniel-montanari replied:
Ok

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+    resp = _user_req_choices(msg, choices).upper()
+    for _ in range(attempts):
+        choice = _choice_from_response(choices, resp)
+        if choice:
```
Correct, I don't think it would make sense to have an empty string as an option.

daniel-montanari replied:
Sounds good

jcollins1983 commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,341 @@
+
+def _user_req_choices(msg: str, choices: tuple):
+    # TODO - do we really need this check since this is a private function and any callers should be calling correctly
+    if len(choices) < 2:
```
Agreed, I've switched to be an assert.

jcollins1983 commented on src/fixate/core/ui.py:
```diff
@@ -55,17 +70,24 @@ def _user_req_choices(msg, choices, target=None, attempts=5):
+    for _ in range(attempts):
+        pub.sendMessage("UI_req_choices", msg=msg, q=q, choices=choices)
+        resp = q.get()
+        ret_val = target(resp, choices)
```
As per my previous comment on this topic.

jcollins1983 replied:
Agree, I don't think it makes much sense for the target to be optional.

jcollins1983 committed:
remove unused import

jcollins1983 commented on src/fixate/ui_cmdline/cmd_line.py:
```diff
@@ -105,14 +107,19 @@ def unregister_cmd_line():
 
 
-def reformat_text(text_str, first_line_fill="", subsequent_line_fill=""):
+def _reformat_text(text_str, first_line_fill="", subsequent_line_fill=""):
     lines = []
+    _wrapper_initial_indent = wrapper.initial_indent
+    _wrapper_subsequent_indent = wrapper.subsequent_indent
     wrapper.initial_indent = first_line_fill
     wrapper.subsequent_indent = subsequent_line_fill
     for ind, line in enumerate(text_str.splitlines()):
         if ind != 0:
             wrapper.initial_indent = subsequent_line_fill
         lines.append(wrapper.fill(line))
+    # reset the indents, calls to this method should not affect the global state
+    wrapper.initial_indent = _wrapper_initial_indent
+    wrapper.subsequent_indent = _wrapper_subsequent_indent
     return "\n".join(lines)
```
What kind of bugs? Not to my knowledge.

jcollins1983 replied:
It's been a while since I looked at this code, but IIRC if you changed the indent at any point, it would stick around on the next call, hence why the global state was being stashed then recalled after modification.
There is usage of the optional arguments at line 171

daniel-montanari replied:
Ah yeah I see now, so you've moved where the indents got reset out of `_user_input` up to here.
We could probably keep simplifying this module to remove the global state but for now the bug fix is fine.

jcollins1983 committed:
switch to consistent use of f-strings

jcollins1983 commented on src/fixate/core/ui.py:
```diff
@@ -55,17 +70,24 @@ def _user_req_choices(msg, choices, target=None, attempts=5):
+    for _ in range(attempts):
+        pub.sendMessage("UI_req_choices", msg=msg, q=q, choices=choices)
+        resp = q.get()
+        ret_val = target(resp, choices)
```
Agree, I don't think it makes much sense for the target to be optional.

jcollins1983 committed:
remove unused import

jcollins1983 committed:
addressing review comments

jcollins1983 committed:
add multiplier to account for width of !

jcollins1983 committed:
add manual tests for UI

jcollins1983 committed:
add file used in my setup

jcollins1983 reviewed (commented):
I started answering comments in VS Code and it decided that I was doing a review... anyhow... here we are.
I think I've addressed all the things you raised. I've left the conversations as unresolved, as I like to leave that to the reviewer to determine. I also added a file that I used for manual testing of how the UI is displayed.

jcollins1983 commented:
This is ready to be reviewed again.

jcollins1983 committed:
update release notes

jcollins1983 committed:
one day I will remember all of these things together... bump version

jcollins1983 committed:
Merge branch 'main' into revamp-ui

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    return _user_choices(msg, CHOICES, attempts)
+
+
+def _user_retry_abort_fail(msg: str, attempts: int = 1) -> str:
```
Should this be private? It isn't used here but is imported in other modules. This highlights a different colour in vs code to show it isn't used and makes me want to delete it.

jcollins1983 replied:
It's private in the sense that we don't want to expose it as part of the public interface. Why Pylance doesn't highlight it as used I don't know, it knows where it's used if you ctrl+click on in or show references though.

I have the same tendency to want to delete things that aren't used, since the mechanism used in `__init__.py` is to expose the methods that _are_ part of the public interface, I've removed the `_` so that  Pylance treats it properly, and updated the comment to indicate that it should not be included in the public interface. You still have to explicitly import it from a private module to get to it. I think I'm happy enough with that approach.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/main.py:
```diff
@@ -11,7 +11,8 @@
 import fixate.config
 from fixate.core.exceptions import SequenceAbort
-from fixate.core.ui import user_info_important, user_serial, user_ok
+from fixate.core.ui import user_info_important, user_ok
```
I'm a bit confused at the difference between core.ui and _ui

daniel-montanari replied:
Probably shouldn't be using core at all

jcollins1983 replied:
Yep, there were missed. Now coming from the new implementation.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    Defines a validator object that can be used to validate user input.
+    """
+
+    def __init__(self, func: Callable[[T], bool], errror_msg: str = "Invalid input"):
```
typo

jcollins1983 replied:
Fixed.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+from collections import OrderedDict
+
+
+T = TypeVar("T")
```
https://typing.python.org/en/latest/spec/generics.html

jcollins1983 replied:
Switched to the alternate method of defining generics. Looks cleaner too IMO.

jcollins1983 replied:
Except now it looks like I will need to bump the version of black to get it through

<img width="617" height="174" alt="Image" src="https://github.com/user-attachments/assets/bb29ce9d-1b2d-4b15-93c2-4f78456f528c" />

jcollins1983 replied:
And now that I've done that _and_ with tox doing its thing (which I didn't run locally before pushing because dumb that's why) the CI run is failing because of other things that the newer version of black wants to introduce. This will be a later problem now.

jcollins1983 replied:
Ok, so this rang a bell from when I was doing the ftdi driver update that is being put into a separate repo under PyFixate (coming soonTM). @clint-lawrence and I spoke about either getting rid of the tox black environment or running pre-commit in tox as well.
I think after speaking to @daniel-montanari about this also, I'm leaning towards just removing the black section from tox. This means that we won't have bulk formatting changes coming in on an ostensibly unrelated PR. If any unchanged files are touched in the future, then these will be caught up with the newer black enforcements.

daniel-montanari replied:
Bumped black and reran https://github.com/PyFixate/Fixate/pull/261
I found this pretty easy to merge into my vmux branch. The only conflicts I had were the config files I had touched and my code that had not been run through the new version of black.

If this ends up being a pain to merge for you then I'm happy to abandon that PR. Really it just means a few extra whitespace diffs when someone touches a file that hasn't been formatted in a while.

daniel-montanari commented on src/fixate/ui_cmdline/cmd_line.py:
```diff
@@ -242,27 +178,39 @@ def _user_display(msg):
+    else:
+        disp_str = "GIF"
+    print("\a")
+    _user_display_important(f"Oh oh, {disp_str} display not supported in command line")
```
Maybe don't need this if we think that showing the link to the image is a valid way of handling this?

jcollins1983 replied:
If we're running a script that should display images in the CLI I want it to be clear that the script should probably be run in the GUI.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+
+
+def _user_req_choices(msg: str, choices: tuple[str, ...]) -> str:
+    assert len(choices) >= 2, "There must be at least 2 choices"
```
Ignore what I said about NATA, it uses user_input to choose a number from a list of options.

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+
+def _choice_from_response(choices: tuple[str, ...], resp: str) -> str | Literal[False]:
+    for choice in choices:
+        if resp.startswith(choice[0]):
```
<img width="409" height="151" alt="Image" src="https://github.com/user-attachments/assets/7a3a9b5b-871f-40f2-a566-169669a4ecbd" />

Is this intentionally different to the old code?

jcollins1983 replied:
Yeah, not sure what I was thinking when I did this. 
With what used to be, you'd get a true response if you responded with Y, YE or YES if the choice was YES, but with what I did you'd get a true response if you responded with the same set but also with any other response starting with Y, like YASS, not really what we want. As I said, not really sure what I was thinking when I did this.

In both cases, if there were multiple choices that started with the same letter, you'd get a hit on the first choice that begins with the entered letter if only one letter was entered, this is only an issue in the CLI, with the GUI you get then entire word for the response. With that in mind, I'll go back to the original logic.

Whilst testing this I found an issue with the retry logic where if an invalid response is entered the user is asked again to input, but if attempts is not overridden to more than 1 this shouldn't happen etc. I've fixed this now too.

daniel-montanari replied:
I think with either case using startswith will result in some edge cases rather than an exact match.
Until then sticking with the old logic but fixing the bug should be good.

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    q: Queue[str] = Queue()
+    pub.sendMessage("UI_block_start")
+    pub.sendMessage("UI_req_choices", msg=msg, q=q, choices=choices)
+    resp = q.get()
```
Does this have the same bug here https://github.com/PyFixate/Fixate/issues/213

jcollins1983 replied:
No, `_user_req_choices` only gets a value from the user (in the GUI shows buttons to press and gets the value for the button that was pressed). It's up to the caller to process that response. Currently the only caller is `_user_choices` which will either get a valid choice from the input or return False (which is what would happen if the user entered 0 or any other falsey value), in which case either another attempt is allowed or an exception is raised.

Currently there is no "gerneric" form of user_choices (the public version of this was removed in 0.5.2 - see line 204 of the release-notes), so there is currently no mechanism for a script to introduce a potentially problematic set of choices, though I have tested with a use of the private `_user_choices` and it happily accepts any string including an empty string and will return that as the choice if that is entered. It currently doesn't accept any numeric values.

Also, I made a point of addressing #213 in this PR.

daniel-montanari replied:
Sounds good.

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    return resp
+
+
+def user_input(msg: str) -> str:
```
Is it intentional that the retry logic has been removed from this? In the old ui user_input and user_input_float are the same thing with a different validator

jcollins1983 replied:
It hasn't, `user_input` from the original `ui` module has no retry logic. It's just used to get a raw input from the user. 

In general, I've made anything that gets user input just that, a mechanism for getting input. If a more specific function uses the function to get input and there is any processing to be done, this is done by the caller, including any retry logic.

daniel-montanari replied:
Isn't there retry logic buried in `_user_req_input`? Or does it not trigger for `user_input` in the `ui` module?

I agree though with moving retry logic up a level rather than having it buried like this.

jcollins1983 replied:
_user_req_input is in the old UI _only_. Keep in mind that the old UI will be phased out. The stuff that really matters is in _ui.py.

daniel-montanari replied:
Yep

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    Raises:
+        UserInputError: If the user fails to enter a number after the specified number of attempts
+    """
+    resp = _user_request_input(msg)
```
Does it make more sense to move resp into the start of the loop and remove from the except

jcollins1983 replied:
Yes, this will fix the bug that I mentioned earlier in `_user_choices` if attempts = 1.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+    Returns:
+        resp (str): The user response from the UI
+    """
+    resp = _user_request_input(msg)
```
Same here

jcollins1983 replied:
Ditto.

daniel-montanari replied:
ok

daniel-montanari commented on src/fixate/_ui.py:
```diff
@@ -0,0 +1,344 @@
+
+
+def _user_choices(msg: str, choices: tuple[str, ...], attempts: int = 5) -> str:
+    resp = _user_req_choices(msg, choices).upper()
```
Same here

jcollins1983 replied:
Yep, found it before I go to this comment. I would much prefer GH comments to be linear with respect to the code instead of temporally linear (they probably would be in a different view... 🤷‍♂️)

daniel-montanari replied:
Yeah, it makes it a pain when running through the code, but has saved me when I split thoughts across messages.
I do need to make comments make sense without needing the context of other comments.

daniel-montanari commented on test/manual/test_ui.py:
```diff
@@ -0,0 +1,107 @@
+PASSING_TESTS = [TestsPassing()]
+FAILING_TESTS = [TestsFailing()]
+
+test_data = {
```
Is there a reason these require an argument and don't play one after the other?

jcollins1983 replied:
Yes, the post sequence info is dependent on the overall result of the test.

daniel-montanari replied:
ok

daniel-montanari reviewed (commented):
Old tests still pass.
New tests seem ok.

A few comments to talk over but nothing majorly wrong.

jcollins1983 committed:
address review comments round 2 and bump black version

jcollins1983 commented:
@daniel-montanari I think I've addressed all of your comments. There were some outdated comments that I marked as resolved, but there are still a few comments where I responded and in general, I prefer the reviewer to mark as resolved if they're happy with the response - so I've left them for you to mark as resolved if you're happy.

jcollins1983 committed:
sync black version to pre-commit...

jcollins1983 committed:
remove black from tox

jcollins1983 committed:
forgot the envlist entry for removing black

jcollins1983 committed:
update mypy version to latest

jcollins1983 committed:
remove tox -e black from workflows

daniel-montanari commented on src/fixate/ui_cmdline/cmd_line.py:
```diff
@@ -125,15 +131,13 @@ def _user_action(msg, callback_obj):
-    :param abort:
-     The queue object to abort this monitoring as the test has already passed.
+    :param callback_obj:
+     The callback object to handle user canncellation or target completion. This is a UserActionCallback object as defined in fixate._ui.user_action.
```
canncellation

jcollins1983 replied:
Fixed

daniel-montanari commented on src/fixate/ui_gui_qt/ui_gui_qt.py:
```diff
@@ -716,7 +721,7 @@ def _topic_UI_action(self, msg, callback_obj):
         :param msg: Information for the user
         :param callback_obj:
-         callback_obj used to communicate back to the user_action call in ui.py.
+         The callback object to handle user canncellation or target completion. This is a UserActionCallback object as defined in fixate._ui.user_action.
```
canncellation

jcollins1983 replied:
Fixed

daniel-montanari commented on .github/workflows/test.yml:
```diff
@@ -32,9 +32,6 @@ jobs:
     - name: Run tests with tox
       run: |
         tox -e py
-    - name: Check formatting
```
Is pre-commit being used now instead of tox for running black?
I can't see any job for formatting in the last CI run.

jcollins1983 replied:
I refer you to this comment @daniel-montanari 
https://github.com/PyFixate/Fixate/pull/214#discussion_r3293861338

daniel-montanari replied:
Replied to that comment

daniel-montanari reviewed (changes requested):
Check CI, minor comments otherwise.
Will need to pull latest changes into this to fix failing tests.

jcollins1983 commented:
> Check CI, minor comments otherwise. Will need to pull latest changes into this to fix failing tests.

What failing tests? And what specifically in the CI? Vague comments don't help in the resolution of these things...

jcollins1983 committed:
fix typos

jcollins1983 committed:
Merge branch 'main' into revamp-ui

jcollins1983 commented:
@daniel-montanari I've addressed the conflict, tests are passing. There are a whole bunch of unresolved comments in outdated sections that I would like you to either provide further comment on or resolve if you're happy.

With regard to the formatting. It is currently reliant upon pre-commit and the person doing any work having the pre-commit hooks installed. I'm open to including pre-commit in the actions to use the same configuration that each person should be using when making changes in a branch. The reason I left it out is that when we need to bump versions of black etc. to accommodate for new features required for the code we're modifying we don't end up with the entire codebase being updated "just because", I would prefer to defer this to when/if this code needs to be modified in the future, but if there's enough desire from others to continue to format the entire codebase each time the formatting tools are updated then I'll acquiesce.

daniel-montanari commented:
> > Check CI, minor comments otherwise. Will need to pull latest changes into this to fix failing tests.
> 
> What failing tests? And what specifically in the CI? Vague comments don't help in the resolution of these things...

I should've been more specific.
There were some tests failing due to breaking changes in packages. They would've only happened if you reran the CI.

daniel-montanari reviewed (commented):
Will approve pending your decision of what to do with the formatting.
Can either have a go merging my changes in or abandon and go with these changes.
https://github.com/PyFixate/Fixate/pull/214#discussion_r3293861338
https://github.com/PyFixate/Fixate/pull/261

jcollins1983 committed:
Merge branch 'main' into revamp-ui

jcollins1983 committed:
black on new files

daniel-montanari reviewed (approved):
Looks good
